File: 2022/src/bin/21.rs
// https://adventofcode.com/2022/day/21

use std::collections::HashMap;

fn create_monkeys(input: &str) -> HashMap<&str, &str> {
    let mut monkeys: HashMap<&str, &str> = HashMap::new();

    input.lines().for_each(|line| {
        let mut parts = line.split(": ");
        let key = parts.next().unwrap();
        let value = parts.next().unwrap();

        monkeys.insert(key, value);
    });

    return monkeys;
}

fn operate(operator: char, x: u64, y: u64) -> u64 {
    match operator {
        '+' => x + y,
        '-' => x - y,
        '*' => x * y,
        '/' => x / y,
        _ => unreachable!(),
    }
}

fn get_value(monkeys: &HashMap<&str, &str>, key: &str) -> u64 {
    let value = monkeys.get(key).unwrap();

    match value.parse::<u16>() {
        // The value is a number
        Ok(number) => number as u64,
        // The value is an equation
        Err(_) => {
            let mut parts = value.split(" ");
            let key1 = parts.next().unwrap();
            let operator = parts.next().unwrap().chars().next().unwrap();
            let key2 = parts.next().unwrap();

            let value1 = get_value(monkeys, key1);
            let value2 = get_value(monkeys, key2);

            operate(operator, value1, value2)
        }
    }
}

fn part1(input: &str) -> u64 {
    let monkeys = create_monkeys(input);

    get_value(&monkeys, "root")
}

fn main() {
    let input = include_str!("../../input/21.txt").trim_end();

    println!("part 1: {}", part1(input));
}

#[cfg(test)]
mod tests {
    const INPUT: &str = include_str!("../../input/21.txt");

    #[test]
    fn part1() {
        assert_eq!(super::part1(INPUT.trim_end()), 121868120894282);
    }
}

File: typescript/2022/21.ts
// https://adventofcode.com/2022/day/21

import { operate } from "../lib/math";

function createMonkeys(input: string): Map<string, number | string[]> {
  const monkeys = new Map<string, number | string[]>();

  input.split("\n").forEach(line => {
    const [_, key, value] = line.match(/(\w{4}): (.+)/)!;

    if (Number.isInteger(parseInt(value, 10))) {
      monkeys.set(key, parseInt(value, 10));
    } else {
      monkeys.set(key, value.split(" "));
    }
  });

  return monkeys;
}

function getValue(monkeys: Map<string, number | string[]>, key: string): number {
  const value = monkeys.get(key)!;

  if (Number.isInteger(value)) {
    return value as number;
  }

  const [key1, operator, key2] = value as string[];

  const value1 = getValue(monkeys, key1);
  const value2 = getValue(monkeys, key2);

  return operate(operator, value1, value2);
}

export function part1(input: string): number {
  const monkeys = createMonkeys(input);

  return getValue(monkeys, "root");
}

const input = (await Bun.file("../_input/2022/21.txt").text()).trimEnd();

if (import.meta.env.NODE_ENV === "test") {
  const { test, expect } = await import("bun:test");

  test("part 1", () => expect(part1(input)).toBe(121868120894282));
} else {
  console.log("part 1:", part1(input));
}
